runtime/arm64.S

```diff
@@ -285,29 +375,69 @@ FUNCTION(caml_allocN)
 
 FUNCTION(caml_c_call)
         CFI_STARTPROC
-    /* Preserve return address in callee-save register x19 */
-        mov     x19, x30
-        CFI_REGISTER(30, 19)
-    /* Record lowest stack address and return address */
-        str     x30, Caml_state(last_return_address)
-        add     TMP, sp, #0
-        str     TMP, Caml_state(bottom_of_stack)
+        stp     x29, x30, [sp, -16]!
+        CFI_ADJUST(16)
+        add     x29, sp, #0
+    /* Switch from OCaml to C */
+        SWITCH_OCAML_TO_C
     /* Make the exception handler alloc ptr available to the C code */
         str     ALLOC_PTR, Caml_state(young_ptr)
-        str     TRAP_PTR, Caml_state(exception_pointer)
+        str     TRAP_PTR, Caml_state(exn_handler)
     /* Call the function */
         blr     ADDITIONAL_ARG
     /* Reload alloc ptr  */
         ldr     ALLOC_PTR, Caml_state(young_ptr)
+    /* Load ocaml stack */
+        SWITCH_C_TO_OCAML
     /* Return */
-        ret     x19
+        ldp     x29, x30, [sp], 16
+        ret
         CFI_ENDPROC
         END_FUNCTION(caml_c_call)
 
+FUNCTION(caml_c_call_stack_args)
+        CFI_STARTPROC
+    /* Arguments:
+        C arguments  : x0 to x7, d0 to d7
+        C function   : ADDITIONAL_ARG
+        C stack args : begin=STACK_ARG_BEGIN
+                       end=STACK_ARG_END */
+        stp     x29, x30, [sp, -16]!
+        CFI_ADJUST(16)
+        add     x29, sp, #0
+    /* Switch from OCaml to C */
+        SWITCH_OCAML_TO_C
+    /* Make the exception handler alloc ptr available to the C code */
+        str     ALLOC_PTR, Caml_state(young_ptr)
+        str     TRAP_PTR, Caml_state(exn_handler)
+    /* Store sp to restore after call */
+        mov     x19, sp
```

Comment:
So the proposed change doesn't work. 

At this point in `caml_c_call_stack_args` we are on the C stack - the previous `stp x29, x30 [sp, -16]!` is done while on the OCaml stack (note the `SWITCH_OCAML_TO_C`). We are storing `sp` in `x19` since the following C call (`blr ADDITIONAL_ARG`) could be allocating and we need to unwind the stack arguments we have pushed to the C-stack on return.

Given the C call here needs us to destroy registers, I don't think using `x19` in this case is so bad?

Reply:
OK, I missed a fine point of stack switching...  I was so happy to have found a use for the frame pointer after all these years of complaining that it just eats space up :-)  

No, destroying x19 is not a big deal.